src/persist-client/src/batch.rs

```diff
@@ -1022,22 +1033,54 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
 
         let (stats, (buf, encode_time)) = isolated_runtime
             .spawn_named(|| "batch::encode_part", async move {
-                let stats = if cfg.stats_collection_enabled {
-                    let stats_start = Instant::now();
-                    match part_stats_for_legacy_part(&schemas, &updates.updates) {
-                        Ok(x) => {
-                            let mut trimmed_bytes = 0;
-                            let x = LazyPartStats::encode(&x, |s| {
-                                trimmed_bytes = trim_to_budget(s, cfg.stats_budget, |s| {
-                                    cfg.stats_untrimmable_columns.should_retain(s)
-                                });
-                            });
-                            Some((x, stats_start.elapsed(), trimmed_bytes))
-                        }
+                // Only encode our updates in a structured format if required, it's expensive.
+                let s = if cfg.stats_collection_enabled || cfg.batch_columnar_format.is_structured()
+                {
+                    let result = metrics_
+                        .columnar
+                        .arrow()
+                        .measure_part_build(|| encode_updates(&schemas, &updates.updates));
+                    match result {
                         Err(err) => {
-                            error!("failed to construct part stats: {}", err);
+                            tracing::error!(?err, "failed to encode in columnar format!");
                             None
                         }
+                        Ok(s) => Some(s),
+                    }
+                } else {
+                    None
+                };
+
+                let stats = if let Some((part, stats)) = s {
+                    // Only change the updates to use the structured format, if our updates had a
+                    // single batch.
+                    match updates.updates {
```

Comment:
No concrete counter-proposal, but I'm having trouble following the control flow of this if/else structure, both exactly what happens in the various conditions and why. For example, feels like this could be behind only `cfg.batch_columnar_format.is_structured()` but it's also gated by `cfg.stats_collection_enabled` (or maybe I'm reading the code wrong! not confidant)

Reply:
Yeah the flow isn't super clear, the tricky bit is if we somehow fail to encode in our columnar format we don't (yet?) want to panic and instead bail on stats collection. I re-worked this using labeled blocks which I think allows it to be a bit cleaner, let me know what you think!